File: neon/Events.lua
--
-- Created by IntelliJ IDEA.
-- User: dantezhu
-- Date: 14-11-7
-- Time: 上午11:51
-- To change this template use File | Settings | File Templates.
--

local M = class("Events")

function M:ctor()
    self._eventTable = {}
end

-- 添加
function M:addHandler(name, callback, target)
    if self._eventTable[name] == nil then
        self._eventTable[name] = {}
    end

    local found = false

    for key, val in pairs(self._eventTable[name]) do
        if val.callback == callback and val.target == target then
            found = true
        end
    end

    if not found then
        table.insert(self._eventTable[name], {
                callback=callback,
                target=target,
        })
    end
end

-- 删除
function M:delHandler(name, callback)
    local handlers = self._eventTable[name]

    if handlers == nil then
        return
    end

    local i = 1
    while i <= #handlers do
        local handler = handlers[i]
        if handler.callback == callback then
            table.remove(handlers, i)
        else
            i = i + 1
        end
    end
end

-- 通过target删除
function M:delHandlersForTarget(target)
    for k, handlers in pairs(self._eventTable) do
        local i = 1
        while i <= #handlers do
            local handler = handlers[i]
            if handler.target == target then
                table.remove(handlers, i)
            else
                i = i + 1
            end
        end
    end
end

-- 删除对应name
function M:delHandlers(name)
    self._eventTable[name] = nil
end

-- 触发
function M:postEvent(name, ...)
    if self._eventTable[name] == nil then
        return
    end

    local handlers = {}
    for i,handler in ipairs(self._eventTable[name]) do
        table.insert(handlers, handler)
    end

    for i, handler in ipairs(handlers) do
        local found = false
        if self._eventTable[name] ~= nil then
            for tmpi, tmpval in ipairs(self._eventTable[name]) do
                if tmpval == handler then
                    found = true
                end
            end
        end

        if found then
            handler.callback(...)
        end
    end
end

return M

File: ts/neon/events.ts
type Callback = (event: string, ...args: any[]) => void;

// Callback的容器
interface Handler {
    callback: Callback,
    target: any,
}

export class Events {
    // 事件及处理函数
    private _eventTable: Map<string, Handler[]>;

    constructor() {
        this._eventTable = new Map<string, Handler[]>();
    }

    // 添加
    addHandler(name: string, callback: Callback, target: any) {
        if (target === null) {
            throw new Error('target cannot be null');
        }

        if (!(this._eventTable.has(name))) {
            this._eventTable.set(name, [])
        }

        let handlers = this._eventTable.get(name) || [];

        let found = false;

        // 判断是否已经存在
        for (let handler of handlers) {
            if (handler.callback === callback && handler.target === target) {
                found = true;
                break;
            }
        }

        if (!found) {
            handlers.push({
                callback: callback,
                target: target,
            })
        }
    }

    // 删除
    delHandler(name: string, callback: Callback) {
        if (!(this._eventTable.has(name))) {
            return;
        }

        let handlers = this._eventTable.get(name) || [];

        for (let i = 0; i < handlers.length; i++) {
            if (handlers[i].callback === callback) {
                // 找到了，删掉
                handlers.splice(i, 1);
                // 因为删除元素后，当前索引位置的元素已经变为下一个元素，所以需要减一
                i--; 
            }
        }

        if (handlers.length === 0) {
            this._eventTable.delete(name);
        }
    }

    // 通过target删除
    delHandlersForTarget(target: any) {
        let keysToDelete:string[] = [];

        this._eventTable.forEach((handlers, name) => {

            for (let i = 0; i < handlers.length; i++) {
                if (handlers[i].target === target) {
                    // 找到了，删掉
                    handlers.splice(i, 1);
                    // 因为删除元素后，当前索引位置的元素已经变为下一个元素，所以需要减一
                    i--;
                }
            }

            if (handlers.length === 0) {
                keysToDelete.push(name);
            }

        });

        if (keysToDelete.length > 0) {
            keysToDelete.forEach(key => {
                this._eventTable.delete(key);
            })
        }
    }

    // 删除对应name
    delHandlers(name: string) {
        this._eventTable.delete(name);
    }

    // 清空
    clearHandlers() {
        this._eventTable.clear()
    }

    // 触发事件
    emit(name: string, ...args: any[]) {
        if (!(this._eventTable.has(name))) {
            return;
        }

        // 浅copy，避免在事件处理函数中添加或删除事件处理函数
        let handlers = (this._eventTable.get(name) || []).slice();

        for (let handler of handlers) {
            if ((this._eventTable.get(name) || []).indexOf(handler) === -1) {
                // 说明可能在过程中被移除了
                // 如果继续调用可能会报错
                continue;
            }

            handler.callback(name, ...args);
        }
    }
}

export const events = new Events();